pre-docs/writing_yaml_plans.md

```diff
@@ -165,91 +142,214 @@ steps:
+
+### Variable references
+
+The simplest way to use a variable is to reference it directly by name.
 
 ```yaml
+parameters:
+  nodes:
+    type: TargetSpec
+
 steps:
-  - name: double_count
-    eval: $count * 2
+  - command: hostname -f
+    target: $nodes
 ```
 
-## Returning results
+This plan takes a parameter called `nodes` and passes it as the target list to a step.
 
-If the `return` key is specified, it will be evaluated after the plan finishes and used as the result of the plan. Otherwise, the plan will return `undef`.
+### String interpolation
 
-The `return` value can be a plain value or it can be a Puppet expression.
+Variables can also be interpolated into string values.
 
-Example:
 ```yaml
+parameters:
+  username:
+    type: String
+
 steps:
-  - name: user_counts
-    task: users::count
+  - task: echo
+    message: "hello ${username}"
     target: $nodes
+```
 
-return: $user_counts.reduce |$sum, $result| { $sum + $result['count'] }
+The string must be double-quoted to allow interpolation.
+
+## Simple expressions
+
+Many operations can be performed on variables to compute new values for step parameters or other fields.
+
+### Array/Hash indexing
+
+You can retrieve a value from an Array or a Hash using the `[]` operator.
+
+This operator can also be used when interpolating a value inside a string.
+
+```yaml
+parameters:
+  users:
+    # Array[String] is a Puppet data type representing an array of strings
+    type: Array[String]
+
+steps:
+  - task: user::add
+    target: 'host.example.com'
+    parameters:
+      name: $users[0]
+  - task: echo
+    target: 'host.example.com'
+    parameters:
+      message: "hello ${users[0]}"
 ```
 
-## Embedding Puppet code expressions
+### Calling functions
 
-You can use a Puppet code expression as the value of any field of a step except the `name`. This allows steps to reference parameters as well as the result of previous steps.
+You can call a Puppet function to compute a value.
 
-Immediately before running a step, any expressions embedded in the step will be evaluated. Each plan parameter and the values of every previous named step will be available in scope.
+See the [Bolt function reference](https://puppet.com/docs/bolt/latest/plan_functions.html) and the [Puppet function reference](https://puppet.com/docs/puppet/latest/function.html) for a list of built-in functions.
 
-Example:
+```yaml
+parameters:
+  users:
+    type: Array[String]
+
+steps:
+  - task: user::add
+    parameters:
+      name: $users.first
+  - task: echo
+    message: "hello ${users.join(',')}"
+```
+
+### Blocks
+
+Some Puppet functions take a block of code as an argument.
+
+For instance, you can filter an array of items based on the result of a block of code.
+
+```yaml
+parameters:
+  numbers:
+    type: Array[Integer]
+
+steps:
+  - task: sum
+    description: "add up the numbers > 5"
+    parameters:
+      indexes: $numbers.filter |$num| { $num > 5 }
```

Comment:
I think we need to explain why this returns an array as opposed to a string briefly here